### ISSUE TITLE
radio@driglu4it: Compatibility to older Cinnamon versions

### ISSUE THREAD

**Issue by NikoKrause** [linuxmint/cinnamon-spices-applets#3374]
There's an issue brought up on the spices website by @Trotter73:

![screenshot-area-2020-12-20-203100](https://user-images.githubusercontent.com/8415124/102722642-be803180-4302-11eb-8890-2431a435caac.png)

@Trotter73, can you tell us, which Cinnamon version are you using?

@jonath92 The command `Util.spawnCommandLineAsyncIO` seems to be 2 years old: https://github.com/linuxmint/cinnamon/blame/master/js/misc/util.js#L212
I guess Cinnamon 4.2+ should be fine.
Versions prior to that, don't have this command.

Can it be replaced by `Util.spawnCommandLineAsyncIO`?

Or should we go multiversion, i.e. offering the `moc` version to Cinnamon 4.0- and the `mpv` version to Cinnamon 4.2+?
I'm fine with both solutions.

**Comment by NikoKrause** [comment]
@jonath92 
I just checked Radio++ on Mint 19.3.
The package `playerctl` seems to be missing in Mint 19.3 / Ubuntu 18.04.
I could install it using the following instructions: https://askubuntu.com/questions/1080671/how-can-i-install-playerctl

There's an issue though. I can start radio streams, but I can't stop them.
So everytime I change the channel, I hear the new station on top of the old ones.

Not sure why it doesn't stop.

**Comment by Trotter73** [comment]
@NikoKrause 

Hi,

Thanks for getting back to me, appreciated !!

```
$ cinnamon --version
Cinnamon 3.8.8
```

Not sure if you have had much to do with the Pi but PiOS is basically Debain, so..

```
lsb_release -a
No LSB modules are available.
Distributor ID:	Raspbian
Description:	Raspbian GNU/Linux 10 (buster)
Release:	10
Codename:	buster
```
 If you need anything else or have any suggestions just shout..

**Comment by NikoKrause** [comment]
@Trotter73 
Thanks for reporting the issue. I've played with the Raspberry Pi a little bit, but I haven't used Cinnamon on it. Didn't know it worked on the Pi :)

I've restored the previous version of the radio applet for Cinnamon versions prior to Cinnamon 4.6: https://github.com/linuxmint/cinnamon-spices-applets/pull/3379
I haven't tested it on Cinnamon 3.8, so it would be nice, if you let me know, if it fixes the radio applet for you.

I think Cinnamon 3.8 is affected by the cache issue as well, which means downloading the applet directly through Cinnamon results in downloading an old version of the applet. Sometimes it can take a few days till the cache reloads.
For now you'll have to download the applet manually on the spices website to get the latest version of the applet: https://cinnamon-spices.linuxmint.com/applets/view/297

**Comment by jonath92** [comment]
Thanks @NikoKrause  for the quick fix. I couldn't find time to answer before but for future developing I want to add some comments: 

> Can it be replaced by Util.spawnCommandLineAsyncIO?
I guess you mean `spawn_async` . This should work I guess

>There's an issue though. I can start radio streams, but I can't stop them.
I have checked this and the reason is that the mpris plugin is not working on LM 19.3 for some reason. I am getting the error: `could not load S0 plugin`. This seems related to: https://github.com/hoyon/mpv-mpris/issues/2. I am not sure as I just flew over the issue but it seems that we would need to compile it ourself to run it. I am not sure if this is an acceptable solution.  Another solution would be to use a different player as using an unofficial plugin is not the most reliable solution anyway but unfortunately that is not that easy. We need a player which can be started headless and with MRPIS Control. VLC was my favorite as it has mpris control and can be stared headless (with `cvlc`) but VLC has a Bug with radio streams in 3.0.9.2 and above (I event tested with nightly build of 4.0). It is showing no metadata. So when you have activated "Show song information on the panel" in the sound applet you are seeing:
![image](https://user-images.githubusercontent.com/49979415/102870022-ba582f00-443c-11eb-9e61-5f4f9eba2cf1.png)
However stopping music and changing the volume is working. And interestingly vlc in Linux Mint 19 is working. In Linux Mint 19 there is a slightly older version of vlc (3.0.8) in the apt repository. Downgrading to an older VLC version in Linux Mint 20 however seems to be not that easily possible.  Another minor disadvantage of using VLC would be that when the radio is running, you can't start radio for the start menu for some reason. 
Another option is Rhythmbox which would have the big advantage that is installed by default. However you can't start Rhytmbox minimized (there is an inofficial plugin but it hasn't been updated for years) and I haven't managed to start a radio stream which is not in my radio playlist, i.e. running this command: 
```
rhythmbox-client --play-uri=https://wdr-wdr2-ruhrgebiet.icecastssl.wdr.de/wdr/wdr2/ruhrgebiet/mp3/128/stream.mp3
```
is only working when you have saved this stream before to your radio playlist. Furthermore the command doesn't give any output in the terminal so we would have no change to improve the error handling. 

So basically I see two options for the future: 
- Using mpv with the mpris plugin for Linux Mint 20 (as long as there is no fix for VLC) and using VLC for older versions
- Keep it as it is (using mpv with the mpris plugin for Linux Mint 20 and moc for older versions)

I am still looking for the perfect audio player for the radio applet though.

**Comment by Trotter73** [comment]
@NikoKrause 

Thank you for coming up with a solution so quickly! With the Pi4s extra memory, USB3 and the KMS driver you can have quite a usable Cinnamon desktop, even with the eye candy on. The Pi4 has slowly replaced my desktop for most things this year, it's just the CPU intensive stuff that is left.

Huge thanks, I will give it a go tonight and report back.
